File: packages/react-native-nitro-image/ios/Image.swift
//
//  Image.swift
//  react-native-nitro-image
//
//  Created by Marc Rousavy on 18.07.24.
//

import Foundation
import UIKit
import NitroModules

class Image : ImageSpec {
  private let uiImage: UIImage
  public var hybridContext = margelo.nitro.HybridContext()
  public var memorySize: Int {
    return getSizeOf(self) + uiImage.memorySize
  }

  public init(uiImage: UIImage) {
    print("✅ Initializing Image...")
    self.uiImage = uiImage
  }

  deinit {
    print("❌ Destroying Image...")
  }

  public var size: ImageSize {
    return ImageSize(uiImage.size.width, uiImage.size.height)
  }

  public var pixelFormat: PixelFormat {
    return .rgb
  }

  var someSettableProp: Double = 1.0

  func toArrayBuffer(format: ImageFormat) throws -> Double {
    throw RuntimeError.error(withMessage: "toArrayBuffer() is not yet implemented!")
  }

  func saveToFile(path: String, onFinished: Func_void_std__string) throws {
    print("Save To File called \(path)...")
    DispatchQueue.main.asyncAfter(deadline: .now() + 5, execute: {
      print("Executing callback now...")
      onFinished(std.string(path))
      print("Callback executed!")
    })
  }
}


extension UIImage {
  var memorySize: Int {
    guard let cgImage else {
      return 0
    }
    return cgImage.bytesPerRow * cgImage.height
  }
}

File: packages/react-native-nitro-image/android/src/main/java/com/margelo/nitro/image/Image.kt
package com.margelo.nitro.image

import android.graphics.Bitmap
import android.os.Handler
import android.os.Looper
import android.util.Log

class Image(val bitmap: Bitmap): HybridImageSpec() {
    override val size: ImageSize
        get() {
            return ImageSize(bitmap.width.toDouble(), bitmap.height.toDouble())
        }
    override val pixelFormat: PixelFormat
        get() = PixelFormat.RGB
    override var someSettableProp: Double
        get() = Math.random()
        set(value) {
            Log.i(TAG, "Set value to $value")
        }

    override fun toArrayBuffer(format: ImageFormat): Double {
        TODO("Not yet implemented")
    }

    override fun saveToFile(path: String, onFinished: (String) -> Unit) {
        Handler(Looper.getMainLooper()).postDelayed({
            onFinished(path)
        }, 5000)
    }

    override val memorySize: Long
        get() = bitmap.allocationByteCount.toLong()
}
